User cart and orders feature

Ayrum1158 opened the pull request.

Commit:
Implemented Identity JWT user registration
Added non generic version of DbResponse

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

Commit:
Merge branch 'APIForProduct' into IdentityJWT

Commit:
Added user login and logout features

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

Commit:
Merge branch 'APIForProduct' into IdentityJWT

Commit:
Added JWT authorisation, tested on TestController
Removed all migration history
Created one migration that contains all previous
(encountered bug too difficult to fix while leaving old migrations)

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

Commit:
Removed unnecessary Logout method in UserService

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

Commit:
Added user cart and orders list

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

Commit:
WIP
Added AddToCart UserController action (doesn't save to db)

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

Commit:
WIP
Working on adding product to user cart

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

Commit:
Reworked FindFirstOrDefaultAsync in GenericRepository to have include features
Fixed AddProductToCart UserRepository method

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

Commit:
Allowed CORS
Added remove product from cart feature

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

Commit:
Added MakeAnOrder feature
Added user orders tables to database
Moved user caer creation to registration method
Removed rudundant code
Removed user cart CreationDate

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

solianyk-sergey commented on InternetStoreCHIReloaded/API/Controllers/ProductsController.cs:
```diff
@@ -27,7 +28,7 @@ public ProductsController(IProductsService productService, IMapper mapper)
         }
 
-        [HttpGet]
+        [HttpGet("Everything")]
```
Should we name it as GetAll?

Ayrum1158 replied:
done

solianyk-sergey commented on InternetStoreCHIReloaded/API/Controllers/UsersController.cs:
```diff
@@ -0,0 +1,112 @@
+
+        [HttpPost]
+        [Authorize]
+        public async Task<GenericResponse> AddToCart(AddToCartViewModel atcViewModel)
```
Please, move it to separate controller

Ayrum1158 replied:
done

solianyk-sergey commented on InternetStoreCHIReloaded/API/Controllers/UsersController.cs:
```diff
@@ -0,0 +1,112 @@
+
+        [HttpPost]
+        [Authorize]
+        public async Task<GenericResponse> RemoveFromCart(RemoveFromCartViewModel rfcViewModel)
```
Please, move it to separate controller

Ayrum1158 replied:
done

solianyk-sergey commented on InternetStoreCHIReloaded/API/Controllers/UsersController.cs:
```diff
@@ -0,0 +1,112 @@
+
+        [HttpPost]
+        [Authorize]
+        public async Task<GenericResponse> MakeAnOrder()
```
Please, move it to separate controller

Ayrum1158 replied:
done

solianyk-sergey commented on InternetStoreCHIReloaded/API/Controllers/UsersController.cs:
```diff
@@ -0,0 +1,112 @@
+            return response;
+        }
+
+        private int GetUserIdFrowJwt()
```
Please, create a BaseController class with this method. All other controllers could be inherited from the BaseController to use this method.
Please, rename method to GetUserId()

Ayrum1158 replied:
done

solianyk-sergey commented on InternetStoreCHIReloaded/API/Startup.cs:
```diff
@@ -67,6 +130,10 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
 
             app.UseRouting();
 
+            app.UseCors(builder => builder.AllowAnyOrigin());
```
Why any origin is allowed?

solianyk-sergey commented on InternetStoreCHIReloaded/BLL/Interfaces/IUsersService.cs:
```diff
@@ -0,0 +1,17 @@
+    {
+        Task<ServiceResult> RegisterUserAsync(UserRegistrationModel newUser);
+        Task<ServiceResult<string>> LoginUserAsync(UserLoggingInModel userModel);
+        Task<ServiceResult> AddToUserCart(int userId, AddToCartModel atcModel);
```
We should have separate service for register/login and cart/order functionality

Ayrum1158 replied:
done

solianyk-sergey replied:
It's not done yet. We still have 1 interface and 1 service. We should have different services.

solianyk-sergey commented on InternetStoreCHIReloaded/BLL/Models/ProductWithQuantity.cs:
```diff
@@ -0,0 +1,15 @@
+
+namespace BLL.Models
+{
+    public class ProductWithQuantity : IHasId
```
ProductWithQuantity is not common name. Is it part of the Cart? If yes, please rename it to something like CartItem

Ayrum1158 replied:
done

solianyk-sergey commented on InternetStoreCHIReloaded/BLL/Models/ProductsSet.cs:
```diff
@@ -0,0 +1,18 @@
+
+namespace BLL.Models
+{
+    public class ProductsSet : IHasId
```
What is ProductsSet? Is it part of the cart? Name is not obvious. please, rename it

Ayrum1158 replied:
done

solianyk-sergey commented on InternetStoreCHIReloaded/BLL/Services/UsersService.cs:
```diff
@@ -0,0 +1,174 @@
+            }
+        }
+
+        public async Task<ServiceResult> AddToUserCart(int userId, AddToCartModel atcModel)// no user validation because we retrieve userId via JWT
```
atc is not common abbreviation. Please, rename the variable

Ayrum1158 replied:
done

solianyk-sergey commented on InternetStoreCHIReloaded/BLL/Services/UsersService.cs:
```diff
@@ -0,0 +1,174 @@
+            return result;
+        }
+
+        public async Task<ServiceResult> RemoveFromUserCart(int userId, RemoveFromCartModel rfcModel)// no user validation because we retrieve userId via JWT
```
Please, rename rfcModel variable

Ayrum1158 replied:
done

solianyk-sergey commented on InternetStoreCHIReloaded/DAL/Entities/OrderedProductEntity.cs:
```diff
@@ -0,0 +1,19 @@
+        public int Quantity { get; set; }
+        public ProductEntity Product { get; set; }
+        public int ProductId { get; set; }
+        public DateTime OrderedDate { get; set; }
```
It could be simplified to Date

Ayrum1158 replied:
done

solianyk-sergey commented on InternetStoreCHIReloaded/DAL/Entities/OrderedProductEntity.cs:
```diff
@@ -0,0 +1,19 @@
+        public ProductEntity Product { get; set; }
+        public int ProductId { get; set; }
+        public DateTime OrderedDate { get; set; }
+        public decimal OrderedPrice { get; set; }
```
It could be simplified to Price

Ayrum1158 replied:
done

solianyk-sergey commented on InternetStoreCHIReloaded/DAL/Migrations/20210812141009_AddedManyToManyBetweenCartEntityAndProductWithEntity.cs:
```diff
@@ -0,0 +1,76 @@
+
+namespace DAL.Migrations
+{
+    public partial class AddedManyToManyBetweenCartEntityAndProductWithEntity : Migration
```
Why relationship between Cart and Cart items is many-to-many?

Ayrum1158 replied:
done

solianyk-sergey commented on InternetStoreCHIReloaded/DAL/Migrations/20210817102603_AddedUserOrders.cs:
```diff
@@ -0,0 +1,125 @@
+                });
+
+            migrationBuilder.CreateTable(
+                name: "OrderEntityUserEntity",
```
Should relationship between Order and User be one-to-many?

Ayrum1158 replied:
One to many*

Ayrum1158 replied:
done

solianyk-sergey commented on InternetStoreCHIReloaded/DAL/Repositories/UsersRepository.cs:
```diff
@@ -0,0 +1,220 @@
+            }
+        }
+
+        public async Task<DbResponse> SetProductToUserCartAsync(int userId, ProductToCartDbModel productToCartDbModel)
```
All business logic from this method should be moved to the Business Logic layer. DAL layer should be as simple as possible and contain queries, insert/update operations and transactions.
Please, move logic to BLL

Ayrum1158 replied:
done

solianyk-sergey commented on InternetStoreCHIReloaded/DAL/Repositories/UsersRepository.cs:
```diff
@@ -0,0 +1,220 @@
+            return dbResponse;
+        }
+
+        public async Task<DbResponse> RemoveProductSetFromUserCartAsync(int userId, int productId)// removes entry from ProductWithQuantity table
```
Please, move all logic to the BLL

Ayrum1158 replied:
done

solianyk-sergey commented on InternetStoreCHIReloaded/DAL/Repositories/UsersRepository.cs:
```diff
@@ -0,0 +1,220 @@
+            return dbResponse;
+        }
+
+        public async Task<DbResponse<int>> GetQuantityOfProductInCartAsync(int userId, int productId)
```
Looks like we don't need those response messages at all. This method could return simple bool result. True - item exists, false - item is absent

solianyk-sergey commented on InternetStoreCHIReloaded/DAL/Repositories/UsersRepository.cs:
```diff
@@ -0,0 +1,220 @@
+            return dbResponse;
+        }
+
+        public async Task<DbResponse> MakeAnOrder(int userId)
```
A lot business logic here. Please, move all business logic to BLL

Ayrum1158 replied:
done

solianyk-sergey commented on InternetStoreCHIReloaded/DAL/Repositories/UsersRepository.cs:
```diff
@@ -0,0 +1,220 @@
+            // getting products only because can't include the product entities in user cart
+            var productIds = user.UserCart.CartItems.Select(i => i.Id).ToList();
+
+            var productsWithQuantity = await _dbcontext.ProductsWithQuantity.Where(p => productIds.Contains(p.Id)).Include(p => p.Product).ToListAsync();
```
Why should we use ProductsWithQuantity? We have this information in the CartItems

Ayrum1158 replied:
done

solianyk-sergey commented on InternetStoreCHIReloaded/DAL/Repositories/UsersRepository.cs:
```diff
@@ -0,0 +1,220 @@
+
+            _dbcontext.ProductsWithQuantity.RemoveRange(productsWithQuantity);
+
+            user.UserCart.CartItems.Clear();
```
Are we going to clear CartItems and ProductsWithQuantity? Are they same?

Ayrum1158 replied:
done

solianyk-sergey commented on InternetStoreCHIReloaded/BLL/Models/Cart.cs:
```diff
@@ -0,0 +1,19 @@
+        public int UserId { get; set; }
+
+        // Cart items:
+        public ICollection<ProductWithQuantity> CartItems { get; set; }
```
It's really not obvious that class and db named ProductWithQuantity when property named CartItems.
Please, rename ProductWithQuantity to CartItems

Ayrum1158 replied:
done

Commit:
Removed unused class ProductsSet
Moved a lot of logic from DAL to BLL
Removed TestController
Moved according logic to according controllers

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

Commit:
Changed several many-to-many to many-to-one relationships

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

Commit:
Changed return type of UsersRepository method

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

solianyk-sergey commented on InternetStoreCHIReloaded/BLL/Services/UsersService.cs:
```diff
@@ -0,0 +1,219 @@
+            }
+        }
+
+        public async Task<ServiceResult> AddToUserCart(int userId, AddToCartModel addToCartModel)// no user validation because we retrieve userId via JWT
```
Please, create separate services for user, cart and order.

solianyk-sergey commented on InternetStoreCHIReloaded/BLL/Services/UsersService.cs:
```diff
@@ -0,0 +1,219 @@
+                };
+            }
+
+            var userEntity = await _usersRepository.GetUserWithCartAsync(userId);
```
Is it required to use _usersRepository? Why can't we use cart repository? Looks like user entity is redundant

solianyk-sergey commented on InternetStoreCHIReloaded/BLL/Services/UsersService.cs:
```diff
@@ -0,0 +1,219 @@
+
+        public async Task<ServiceResult> MakeAnOrder(int userId)// no user validation because we retrieve userId via JWT
+        {
+            var dbResponse = await _usersRepository.MakeAnOrder(userId);
```
Please move all business logic to BLL

solianyk-sergey commented on InternetStoreCHIReloaded/DAL/Repositories/UsersRepository.cs:
```diff
@@ -0,0 +1,209 @@
+            return dbResponse;
+        }
+
+        public async Task<int?> GetQuantityOfProductInCartAsync(int userId, int productId)
```
This method is not optimal. We need to get only product Quantity. We shouldn't fetch user

solianyk-sergey commented on InternetStoreCHIReloaded/BLL/Services/UsersService.cs:
```diff
@@ -0,0 +1,219 @@
+
+            var dbModel = _mapper.Map<ProductToCartDbModel>(addToCartModel);
+
+            var quantityDbResponse = await _usersRepository.GetQuantityOfProductInCartAsync(userId, addToCartModel.ProductId);
```
Looks like we already fetched userCart with cart items. Why should we get from the database cart one more time?

solianyk-sergey commented on InternetStoreCHIReloaded/DAL/Repositories/UsersRepository.cs:
```diff
@@ -0,0 +1,209 @@
+            }
+        }
+
+        public async Task<DbResponse> SetProductToUserCartAsync(int userId, ProductToCartDbModel productToCartDbModel, bool userHasProductInCart)
```
Please, move all logic to the BLL

Ayrum1158 committed:
Separated controllers, services and repositories
Moved most business logic from DAL to BLL

Signed-off-by: Oleksandr Korotetskyi <ayrum1158@gmail.com>

Ayrum1158 committed:
Refactored repositories and Services
Reduced amount of requests to database

Signed-off-by: Oleksandr Korotetskyi <ayrum1158@gmail.com>

Ayrum1158 committed:
Added business rules restrictions

Signed-off-by: Oleksandr Korotetskyi <ayrum1158@gmail.com>

solianyk-sergey reviewed (approved):
LGTM
